django/contrib/auth/hashers.py

```diff
@@ -704,3 +704,70 @@ def safe_summary(self, encoded):
 
     def harden_runtime(self, password, encoded):
         pass
+
+
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
```

Comment:
Increasing this parameters slightly may lead to exceeding the default 32-MiB memory limit.

On my machine setting `work_factor` to 32768 (the next valid value after 16384) results in `ValueError: Invalid parameter combination for n, r, p, maxmem.`.

Could you please make it possible to increase the memory limit?

Reply:
I added maxmem class attribute to allow for memory limit increase in the most recent commit.

Reply:
Thanks!

I see that you included `maxmem` in a string returned by `encode`. Is there any reason it should be there?
`maxmem` should not affect a hash

Reply:
ilia-v,

The work factor limitation you've run into is a limitation of the underlying Python standard library.

```~$ python3
Python 3.9.1 (default, Feb  3 2021, 07:04:15) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from hashlib import scrypt
>>> scrypt(b'password', salt=b'salt', n=16384, r=8, p=1)
b'tW1\xafD\x84\xf3#\x96\x89i\xed\xa2\x89\xae\xee\x00[Y\x03\xacV\x1ed\xa5\xac\xa1!y{\xf7sN\xf9\xfdXB.."\x18;\xca\xcb\xa9\xec\x87\xba\x0c\x83\xb7\xa2\xe7\x88\xf0<\xe0\xda\x06F43\xcd\xa6'
>>> scrypt(b'password', salt=b'salt', n=32768, r=8, p=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Invalid parameter combination for n, r, p, maxmem.```

I agree that it's a problem, but I think fixing it is outside the scope of Django's responsibility. If you want it fixed take it up with the Python project.

Reply:
@kerkeslager `hashlib.scrypt` is a low-level function that takes multiple arguments.

`maxmem` is one of them, it has a default value that does not suit all cases. Therefore, Django should allow changing the value. @ryowright has already made the necessary change.

Note, `maxmem` does not affect generated hashes, its value should not be included in a string saved to a database and should not be compared in `must_update`.

Reply:
@illia-v Agreed, I removed `maxmem` from an encoded password.